datafusion/functions/src/datetime/to_local_time.rs

```diff
@@ -0,0 +1,601 @@
+/// while keep the display value of the timestamp the same.
+///
+/// # Example 1
+///
+/// ```
+/// # use datafusion_common::ScalarValue;
+/// # use datafusion_expr::ColumnarValue;
+/// # use datafusion_functions::datetime::to_local_time::ToLocalTimeFunc;
+/// # use datafusion_expr::ScalarUDFImpl;
+///
+/// // 2019-03-31 01:00:00 +01:00
+/// let res = ToLocalTimeFunc::new()
+///     .invoke(&[ColumnarValue::Scalar(ScalarValue::TimestampSecond(
+///         Some(1_553_990_400),
+///         Some("Europe/Brussels".into()),
+///     ))])
+///     .unwrap();
+///
+/// // 2019-03-31 01:00:00 <-- this timestamp no longer has +01:00 offset
+/// let expected = ScalarValue::TimestampSecond(Some(1_553_994_000), None);
+///
+/// match res {
+///   ColumnarValue::Scalar(res) => {
+///       assert_eq!(res, expected);
+///   }
+///   _ => panic!("unexpected return type"),
+/// }
+/// ```
+///
+/// # Example 2
+///
+/// ```
+/// # use datafusion_common::ScalarValue;
+/// # use datafusion_expr::ColumnarValue;
+/// # use chrono::NaiveDateTime;
+/// # use datafusion_functions::datetime::to_local_time::ToLocalTimeFunc;
+/// # use datafusion_expr::ScalarUDFImpl;
+///
+/// let timestamp_str = "2020-03-31T13:40:00";
+/// let timezone_str = "America/New_York";
+/// let tz: arrow::array::timezone::Tz =
+///     timezone_str.parse().expect("Invalid timezone");
+///
+/// let timestamp = timestamp_str
+///     .parse::<NaiveDateTime>()
+///     .unwrap()
+///     .and_local_timezone(tz) // this is in a local timezone
+///     .unwrap()
+///     .timestamp_nanos_opt()
+///     .unwrap();
+///
+/// let expected_timestamp = timestamp_str
+///     .parse::<NaiveDateTime>()
+///     .unwrap()
+///     .and_utc() // this is in UTC
+///     .timestamp_nanos_opt()
+///     .unwrap();
+///
+/// let input =
+///     ScalarValue::TimestampNanosecond(Some(timestamp), Some(timezone_str.into()));
+/// let res = ToLocalTimeFunc::new()
+///     .invoke(&[ColumnarValue::Scalar(input)])
+///     .unwrap();
+/// let expected = ScalarValue::TimestampNanosecond(Some(expected_timestamp), None);
+/// match res {
+///     ColumnarValue::Scalar(res) => {
+///         assert_eq!(res, expected);
+///         }
+///     _ => panic!("unexpected return type"),
+/// }
+/// ```
+#[derive(Debug)]
+pub struct ToLocalTimeFunc {
+    signature: Signature,
+}
+
+impl Default for ToLocalTimeFunc {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ToLocalTimeFunc {
+    pub fn new() -> Self {
+        let base_sig = |array_type: TimeUnit| {
+            vec![
+                Exact(vec![Timestamp(array_type, None)]),
+                Exact(vec![Timestamp(array_type, Some(TIMEZONE_WILDCARD.into()))]),
+            ]
+        };
+
+        let full_sig = [Nanosecond, Microsecond, Millisecond, Second]
+            .into_iter()
+            .map(base_sig)
+            .collect::<Vec<_>>()
+            .concat();
```

Comment:
TIL https://doc.rust-lang.org/std/vec/struct.Vec.html#method.concat

I think you could potentially make this slightly more efficient using `flat_map` and save collecting an intermediate Vecs:

```suggestion
        let full_sig = [Nanosecond, Microsecond, Millisecond, Second]
            .into_iter()
            .flat_map(base_sig)
            .collect::<Vec<_>>()
```

You could then also change the `closure` above to return a `[]` rather than `vec![]`

```rust
  [
                Exact(vec![Timestamp(array_type, None)]),
                Exact(vec![Timestamp(array_type, Some(TIMEZONE_WILDCARD.into()))]),
            ]
```

```rust
  vec![
                Exact(vec![Timestamp(array_type, None)]),
                Exact(vec![Timestamp(array_type, Some(TIMEZONE_WILDCARD.into()))]),
            ]
```

I don't think this will make a difference performance wise in practice as this struct be created once and then reused many times (via an `Arc<ScalarUDF>`)

Reply:
Neat! Updated in https://github.com/apache/datafusion/pull/11347/commits/5fe18d5e12c2479bfeaf0ce6cc84cdcf7670c93d